### PR TITLE
Add custom tablename support

### PR DESCRIPTION
**Pull request by CyJaySong**
Add custom tablename support

### CODE REVIEW

**Comment by hsluoyz** [comment]
@nodece @GopherJ @dovics @divy9881 please review.

File: adapter.go [hunk 1]
```diff
@@ -25,14 +25,22 @@ import (
...
-	V4    string `xorm:"varchar(100) index not null default ''"`
-	V5    string `xorm:"varchar(100) index not null default ''"`
+	PType   string  `xorm:"varchar(100) index not null default ''"`
+	V0      string  `xorm:"varchar(100) index not null default ''"`
+	V1      string  `xorm:"varchar(100) index not null default ''"`
+	V2      string  `xorm:"varchar(100) index not null default ''"`
+	V3      string  `xorm:"varchar(100) index not null default ''"`
+	V4      string  `xorm:"varchar(100) index not null default ''"`
+	V5      string  `xorm:"varchar(100) index not null default ''"`
+	adapter Adapter `xorm:"-" json:"-"`
```
Comment (nodece) [inline adapter.go:43]: It looks unnecessary.

**Comment by nodece** [comment]
@CyJaySong please add the `NewAdapterWithTableName` method.

**Comment by nodece** [comment]
When you call the `NewAdapter`, casbin table has been created.

**Comment by dovics** [comment]
I don't like to use Adapter as the field of casbinRule, It looks really strange. Perhaps a global variable would be better, what do you think @CyJaySong @hsluoyz @nodece

**Comment by nodece** [comment]
ok great, add a global method named is `SetTableName`.

**Comment by hsluoyz** [comment]
> I don't like to use Adapter as the field of casbinRule, It looks really strange. Perhaps a global variable would be better, what do you think @CyJaySong @hsluoyz @nodece

I agree. It's really weird to have adapter there.

**Comment by hsluoyz** [comment]
@CyJaySong can you make changes based on review comments?

**Comment by CyJaySong** [comment]
I was negligent. I will modify it

Zixuan Liu <notifications@github.com> 于2020年7月9日周四 下午10:35写道：

**Comment by CyJaySong** [comment]
It is not appropriate to use global variables。There is a problem using more
than one xorm adapter

dovics <notifications@github.com> 于2020年7月9日周四 下午10:53写道：

**Comment by nodece** [comment]
@CyJaySong Could you add tests for the feature?

**Comment by CyJaySong** [comment]
> @CyJaySong Could you add tests for the feature?

Is there anything wrong with that?

**Comment by CyJaySong** [comment]
> @CyJaySong Could you add tests for the feature?

Is there anything wrong with that?

**Comment by nodece** [comment]
Looks good,  but we also need to write test case to test the feature.

**Comment by CyJaySong** [comment]
I have no inspiration

**Comment by CyJaySong** [comment]
> Looks good, but we also need to write test case to test the feature.

I have no inspiration

**Comment by hsluoyz** [comment]
`tableName` should not be put into Casbin rule. I think the code quality of this PR is not ideal. Sorry that I have to close it. Feel free to raise a new good one.

**Comment by nodece** [comment]
Hi @hsluoyz , The `tableName` will not be added to DB, due to we add `xorm: "-"` to struct tag, so the implementation is correct.

We can't define a global variable for set table name.
> It is not appropriate to use global variables。There is a problem using more
than one xorm adapter

**Comment by hsluoyz** [comment]
OK. Go ahead!

**Comment by CyJaySong** [comment]
> OK. Go ahead!

So, do you agree to merge 😅

**Comment by hsluoyz** [comment]
> > OK. Go ahead!
> 
> So, do you agree to merge 😅

@nodece can decide.

**Comment by nodece** [comment]
@hsluoyz Could you give me the repository permission?

**Comment by hsluoyz** [comment]
OK. Granted.

**Pull request merged**